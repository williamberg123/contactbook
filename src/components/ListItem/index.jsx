import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';

import StyledListItem from './styles';

export default function ListItem({ contactInfo }) {
	const { firstName, lastName, email, phoneNumber } = contactInfo;

	return (
		<StyledListItem>
			<span>{firstName}</span>
			<span>{lastName}</span>
			<span>{email}</span>
			<span>{phoneNumber}</span>
			<Link to={`/editcontact?id=${email}`}>Editar</Link>
			<Link to="/deletecontact">Excluir</Link>
		</StyledListItem>
	);
}

ListItem.propTypes = {
	contactInfo: PropTypes.shape({
		firstName: PropTypes.string.isRequired,
		lastName: PropTypes.string.isRequired,
		email: PropTypes.string.isRequired,
		phoneNumber: PropTypes.string.isRequired,
	}).isRequired,
};
