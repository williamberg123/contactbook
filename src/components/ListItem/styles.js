import styled from 'styled-components';

const StyledListItem = styled.div`
	padding: 20px;
	width: 100%;
	display: flex;
	align-items: center;
	justify-content: space-between;
	border-top: 1px solid #afafaf;

	& a {
		text-decoration: none;
		padding: 10px;
		color: white;
		border-radius: 3px;
		background-color: #1cb7ff;
	}

	& a:last-child {
		background-color: red;
	}

	@media (max-width: 650px) {
		padding: 20px 0;
		font-size: 0.7rem;

		& a {
			padding: 5px;
		}
	}
`;

export default StyledListItem;
