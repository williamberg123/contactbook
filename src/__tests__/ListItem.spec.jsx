import { fireEvent, render, screen } from '@testing-library/react';
import { describe, expect, it, vi } from 'vitest';
import { BrowserRouter } from 'react-router-dom';
import PropTypes from 'prop-types';

import ListItem from '../components/ListItem';
import AppContext from '../contexts/AppProvider/AppContext';

const context = {
	user: {
		uid: 'oifrtjrr3298rdh34f45',
	},
	deleteContact: vi.fn(),
};

const AppProvider = ({ children }) => {
	return (
		<AppContext.Provider value={context}>{children}</AppContext.Provider>
	);
};

AppProvider.propTypes = {
	children: PropTypes.node,
};

describe('<ListItem />', () => {
	const contactInfo = {
		firstName: 'Edivar',
		lastName: 'Diogo',
		email: 'edivardiogo@gmail.com',
		phoneNumber: '8899999-9999',
		docRef: '8h4f89ug54u56y32r32f43',
	};

	it('should render correctly ListItem', () => {
		expect.assertions(6);
		render(
			<BrowserRouter>
				<AppProvider>
					<ListItem {...contactInfo} />
				</AppProvider>
			</BrowserRouter>,
		);

		expect(screen.getByText(contactInfo.firstName)).toBeInTheDocument();
		expect(screen.getByText(contactInfo.lastName)).toBeInTheDocument();
		expect(screen.getByText(/edivardiogo@gmail.com/i)).toBeInTheDocument();
		expect(screen.getByText(/8899999-9999/i)).toBeInTheDocument();
		expect(screen.getByRole('link', { name: /editar/i })).toBeInTheDocument();
		expect(screen.getByRole('link', { name: /excluir/i })).toBeInTheDocument();
	});

	it('should render links with the correct value to href attribute', () => {
		render(
			<BrowserRouter>
				<AppProvider>
					<ListItem {...contactInfo} />
				</AppProvider>
			</BrowserRouter>,
		);

		expect(screen.getByRole('link', { name: /editar/i })).toHaveAttribute('href', `/editcontact?id=${contactInfo.docRef}`);
		expect(screen.getByRole('link', { name: /excluir/i })).toHaveAttribute('href', '/deletecontact');
	});

	it('should call function when Link is clicked', () => {
		render(
			<BrowserRouter>
				<AppProvider>
					<ListItem {...contactInfo} />
				</AppProvider>
			</BrowserRouter>,
		);

		fireEvent.click(screen.getByRole('link', { name: /excluir/i }));

		expect(context.deleteContact).toBeCalled();
	});

	it('should match snapshot', () => {
		render(
			<BrowserRouter>
				<AppProvider>
					<ListItem {...contactInfo} />
				</AppProvider>
			</BrowserRouter>,
		);

		expect(screen.getByRole('link', { name: /editar/i }).parentElement).toMatchSnapshot();
	});
});
