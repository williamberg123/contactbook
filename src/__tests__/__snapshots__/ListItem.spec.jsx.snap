// Vitest Snapshot v1

exports[`<ListItem /> > should match snapshot 1`] = `
.c0 {
  padding: 20px;
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  border-top: 1px solid #afafaf;
}

.c0 a {
  -webkit-text-decoration: none;
  text-decoration: none;
  padding: 10px;
  color: white;
  border-radius: 3px;
  background-color: #1cb7ff;
}

.c0 a:last-child {
  background-color: red;
}

@media (max-width:650px) {
  .c0 {
    padding: 20px 0;
    font-size: 0.7rem;
  }

  .c0 a {
    padding: 5px;
  }
}

<div
  class="c0"
>
  <span>
    Edivar
  </span>
  <span>
    Diogo
  </span>
  <span>
    edivardiogo@gmail.com
  </span>
  <span>
    8899999-9999
  </span>
  <a
    href="/editcontact?id=8h4f89ug54u56y32r32f43"
  >
    Editar
  </a>
  <a
    href="/deletecontact"
  >
    Excluir
  </a>
</div>
`;
